### PR TITLE
Consistency Checker dev branch to master 

### PR DESCRIPTION
**Pull request by ao508**
Consistency checker application subscribes to two topics. One more incoming requests directly
from LIMS and another for requests to check for consistency coming from MetaDB. The consistency checker
compares these two request JSONs and returns a boolean if passed or failed check.

Message handling service:
- Added message handling service
- Added handler for both subscribed topics
  1. subscriber for topic "igo.new-request"
  2. subscriber for topic "metadb.new-request-consistency-check"

Consistency checker logging:
- Added logic for determining whether requests were processed/handled within
  expected messaging time threshold (default is 300 seconds)

Unit testing:
- Added mock request jsons covering (1) incoming requests, (2) published requests,
  and (3) malformed request cases
- Tests added for checking consistency between incoming json and published json
- Tests added for checking requests with null/empty fields

Co-authored-by: Angelica Ochoa <15623749+ao508@users.noreply.github.com>
Co-authored-by: Divya Madala <71040191+divyamadala30@users.noreply.github.com>

Signed-off-by: Angelica Ochoa <ochoaa@mskcc.org>

### CODE REVIEW

File: src/main/java/org/mskcc/cmo/metadb/service/impl/MessageHandlingServiceImpl.java [hunk 1]
```diff
@@ -0,0 +1,361 @@
...
+                        //    request.setTopic(CMO_NEW_REQUEST_TOPIC)  if the time exceeded to publish
+                        //}
+
+                        // remove request from consistency check messages received
+                        System.out.println("Removing request from consistency check messages received: " + request.getRequestId());
+                        metadbRequestConsistencyCheckerMessagesReceived.remove(request.getRequestId());
+
+                        // save request details to logger file
+                        // request details should have everything now..
+                        File loggerFile = fileUtil.getOrCreateFileWithHeader(consistencyCheckerFailuresFilepath,
```
Comment (n1zea144) [inline src/main/java/org/mskcc/cmo/metadb/service/impl/MessageHandlingServiceImpl.java:287]: Should we move this to the constructor of NewIgoRequestHandler?

File: src/main/java/org/mskcc/cmo/metadb/service/impl/MessageHandlingServiceImpl.java [hunk 2]
```diff
@@ -0,0 +1,361 @@
...
+                        if (passedConsistencyCheck) {
+                            System.out.println("Passed consistency check, marking request as passed and adding to request publishing queue");
+                            requestsToCheck.setStatusType(ConsistencyCheckerRequest.StatusType.SUCCESSFUL);
+                            // only add request to publishing queue if it passed the consistency check
+                            requestPublishingQueue.add(requestsToCheck);
+                        } else {
+                            System.out.println("request FAILED consistency check: " + requestsToCheck.getRequestId());
+                            requestsToCheck.setStatusType(ConsistencyCheckerRequest.StatusType.FAILED_INCONSISTENT_REQUEST_JSONS);
+                            // save details to publishing failure logger
+                            File loggerFile = fileUtil.getOrCreateFileWithHeader(consistencyCheckerFailuresFilepath,
```
Comment (n1zea144) [inline src/main/java/org/mskcc/cmo/metadb/service/impl/MessageHandlingServiceImpl.java:342]: Should we move this to the constructor of ConsistencyCheckerHandler?

File: src/main/java/org/mskcc/cmo/metadb/service/impl/MessageHandlingServiceImpl.java [hunk 3]
```diff
@@ -0,0 +1,361 @@
...
+                // if so then assume dropped?
+                // if status is already set as StatusType.FAILED_DROPPED_MESSAGE then assume
+                // we already know that the message has been dropped, no need to log it again
+                if (igoNewRequest.getStatusType().equals(ConsistencyCheckerRequest.StatusType.FAILED_DROPPED_MESSAGE)) {
+                    continue;
+                }
+                igoNewRequest.setStatusType(ConsistencyCheckerRequest.StatusType.FAILED_DROPPED_MESSAGE);
+                igoNewRequestMessagesReceived.put(incomingRequestId, igoNewRequest); // update key-value in map
+                try {
+                    File loggerFile = fileUtil.getOrCreateFileWithHeader(consistencyCheckerFailuresFilepath,
```
Comment (n1zea144) [inline src/main/java/org/mskcc/cmo/metadb/service/impl/MessageHandlingServiceImpl.java:226]: We should pull this out of the loop.

File: src/main/java/org/mskcc/cmo/metadb/service/impl/MessageHandlingServiceImpl.java [hunk 4]
```diff
@@ -0,0 +1,361 @@
...
+            ConsistencyCheckerRequest igoNewRequest = igoNewRequestMessagesReceived.get(incomingRequestId);
+            if (metadbRequestConsistencyCheckerMessagesReceived.containsKey(incomingRequestId)) {
+                // add to consistency checking queue if request is in both sets of
+                // messages received
+                System.out.println("Found same request in both queues: " + incomingRequestId);
+
+                // not removing from map because we want to use the incoming timestamp to determine
+                // whether message took longer than expected to receive from CMO_NEW_REQUEST_CONSISTENCY_CHECKER
+                // this request will be removed when it is published to CMO_NEW_REQUEST
+                ConsistencyCheckerRequest metadbConsistencyCheckRequest = metadbRequestConsistencyCheckerMessagesReceived.get(incomingRequestId);
```
Comment (n1zea144) [inline src/main/java/org/mskcc/cmo/metadb/service/impl/MessageHandlingServiceImpl.java:206]: I understand this has to do with a message that arrives, but arrives late.  I'm not following this section of code, lets discuss over zoom.

**Review by n1zea144 — COMMENTED** [review]
I like all the unit tests - I've left comments, mostly in the messaging handling code.  We can discuss over zoom.

File: src/main/resources/log4j.properties.EXAMPLE [hunk 5]
```diff
@@ -0,0 +1,15 @@
+# Change INFO to DEBUG, if you want to see debugging info on underlying libraries we use.
```
[the remark this answers is not in the document]
Comment (ao508) [reply src/main/resources/log4j.properties.EXAMPLE:1]: done

File: src/main/java/org/mskcc/cmo/metadb/model/ConsistencyCheckerRequest.java [hunk 6]
```diff
@@ -0,0 +1,134 @@
...
+import org.apache.commons.lang3.StringUtils;
+
+/**
+ *
+ * @author ochoaa
+ */
+public class ConsistencyCheckerRequest {
+    public enum StatusType {
+        SUCCESSFUL,
+        SUCCESSFUL_PUBLISHING_TIME_EXCEEDED,
```
[the remark this answers is not in the document]
Comment (ao508) [reply src/main/java/org/mskcc/cmo/metadb/model/ConsistencyCheckerRequest.java:29]: done

File: src/main/java/org/mskcc/cmo/metadb/ConsistencyCheckerApp.java [hunk 7]
```diff
@@ -0,0 +1,60 @@
...
+
+    @Autowired
+    private MessageHandlingService messageHandlingService;
+
+    private Thread shutdownHook;
+    final CountDownLatch consistencyCheckerAppClose = new CountDownLatch(1);
+
+    @Override
+    public void run(String... args) throws Exception {
+        System.out.println("Starting up MetaDB application...");
```
[the remark this answers is not in the document]
Comment (ao508) [reply src/main/java/org/mskcc/cmo/metadb/ConsistencyCheckerApp.java]: done

File: src/main/resources/application.properties.EXAMPLE [hunk 8]
```diff
@@ -0,0 +1,25 @@
...
+igo.new_request_topic=
+metadb.new_request_consistency_check_topic=
+metadb.cmo_new_request_topic=
+
+# threading
+num.new_request_handler_threads=
+num.consistency_checker_handler_threads=
+
+# messaging time threshold in seconds (default is 5 minutes)
+messaging.time_threshold_seconds=300
```
Comment (n1zea144) [inline src/main/resources/application.properties.EXAMPLE:19]: This default is already set in the code, should we leave it blank here?

File: src/main/java/org/mskcc/cmo/metadb/service/impl/MessageHandlingServiceImpl.java [hunk 9]
```diff
@@ -0,0 +1,397 @@
...
+    private static boolean initialized = false;
+    private static volatile boolean shutdownInitiated;
+    private static final ExecutorService exec = Executors.newCachedThreadPool();
+    private final ObjectMapper mapper = new ObjectMapper();
+
+    @Override
+    public void initialize(Gateway gateway) throws Exception {
+        if (!initialized) {
+            messagingGateway = gateway;
+            setupConsistencyCheckerHandler(messagingGateway, this);
```
Comment (n1zea144) [inline src/main/java/org/mskcc/cmo/metadb/service/impl/MessageHandlingServiceImpl.java:94]: Can we rename this to setupConsistencyCheckerSubscriber?

File: src/main/java/org/mskcc/cmo/metadb/service/impl/MessageHandlingServiceImpl.java [hunk 10]
```diff
@@ -0,0 +1,397 @@
...
+            setupIgoNewRequestHandler(messagingGateway, this);
```
Comment (n1zea144) [inline src/main/java/org/mskcc/cmo/metadb/service/impl/MessageHandlingServiceImpl.java:95]: Can we rename this to setupIgoNewRequestSubscriber?

**Review by n1zea144 — CHANGES_REQUESTED** [review]

File: src/main/java/org/mskcc/cmo/metadb/util/ConsistencyCheckerUtil.java [hunk 11]
```diff
@@ -0,0 +1,187 @@
...
+        }
+        return assertResponse;
+    }
+
+    public Boolean isConsistent(String referenceJson, String targetJson) throws Exception {
+        return isConsistent(referenceJson, targetJson, DEFAULT_IGNORED_FIELDS);
+    }
+
+    /**
+     * TODO?
```
[the remark this answers is not in the document]
Comment (ao508) [reply src/main/java/org/mskcc/cmo/metadb/util/ConsistencyCheckerUtil.java:78]: i might just remove this for now

**Comment by ao508** [comment]
@n1zea144 requested changes have been made

File: src/main/java/org/mskcc/cmo/metadb/model/ConsistencyCheckerRequest.java [hunk 12]
```diff
@@ -0,0 +1,178 @@
+package org.mskcc.cmo.metadb.model;
+
+import java.util.Arrays;
+import java.util.List;
+import org.apache.commons.lang3.StringUtils;
+
+/**
+ *
+ * @author ochoaa
```
Comment (ao508) [inline src/main/java/org/mskcc/cmo/metadb/model/ConsistencyCheckerRequest.java:9]: add comment describing class

File: src/main/java/org/mskcc/cmo/metadb/service/impl/MessageHandlingServiceImpl.java [hunk 13]
```diff
@@ -0,0 +1,397 @@
...
+                            request.setStatusType(StatusType.SUCCESSFUL_PUBLISHING_TIME_EXCEEDED);
+                        }
+
+                        // remove request from consistency check messages received
+                        LOG.debug("Removing request from consistency check messages received: "
+                                + request.getRequestId());
+                        consistencyCheckerMessagesReceived.remove(request.getRequestId());
+
+                        // save request details to logger file
+                        File loggerFile = fileUtil.getOrCreateFileWithHeader(
```
Comment (n1zea144) [inline src/main/java/org/mskcc/cmo/metadb/service/impl/MessageHandlingServiceImpl.java:317]: Lets pull this file open code out of this loop.

File: src/main/java/org/mskcc/cmo/metadb/service/impl/MessageHandlingServiceImpl.java [hunk 14]
```diff
@@ -0,0 +1,397 @@
...
+                            // only add request to publishing queue if it passed the consistency check
+                            requestPublishingQueue.add(requestsToCheck);
+                        } else {
+                            LOG.warn("Request failed consistency check: " + requestsToCheck.getRequestId()
+                                    + ", storing details to consistency check failures log file");
+                            requestsToCheck.setStatusType(
+                                    ConsistencyCheckerRequest.StatusType.FAILED_INCONSISTENT_REQUEST_JSONS);
+
+                            // save details to publishing failure logger
+                            File loggerFile = fileUtil.getOrCreateFileWithHeader(
```
Comment (n1zea144) [inline src/main/java/org/mskcc/cmo/metadb/service/impl/MessageHandlingServiceImpl.java:378]: Lets pull this file open stuff out of this loop.

File: src/main/java/org/mskcc/cmo/metadb/service/impl/MessageHandlingServiceImpl.java [hunk 15]
```diff
@@ -0,0 +1,399 @@
...
+
+    @Autowired
+    private ConsistencyCheckerUtil consistencyCheckerUtil;
+
+    private File loggerFile;
+
+    @Autowired
+    private void initFileUtilLogger() throws Exception {
+        ConsistencyCheckerRequest header = new ConsistencyCheckerRequest();
+        this.loggerFile = fileUtil.getOrCreateFileWithHeader(
```
Comment (n1zea144) [inline src/main/java/org/mskcc/cmo/metadb/service/impl/MessageHandlingServiceImpl.java:77]: I didn't pick up on this until you made these changes, but do you know if BufferedWriter is thread safe?  That is, you have two threads writing to the same file, potentially at the same time.  My guess is that it is, but I would confirm.  You probably should put a call to flush here:
https://github.com/mskcc/cmo-metadb-common/blob/master/src/main/java/org/mskcc/cmo/common/impl/FileUtilImpl.java#L46

**Review by n1zea144 — CHANGES_REQUESTED** [review]

**Pull request merged**